libevmasm/ConstantOptimiser.cpp

```diff
@@ -185,6 +192,66 @@ AssemblyItems const& CodeCopyMethod::copyRoutine()
 	return copyRoutine;
 }
 
+bigint ShortCodeCopyMethod::gasNeeded() const
+{
+	return combineGas(
+		// Run gas: we ignore memory increase costs
+		simpleRunGas(copyRoutine()) + GasCosts::copyGas,
+		// Data gas for copy routines: Some bytes are zero, but we ignore them.
+		bytesRequired(copyRoutine()) * (m_params.isCreation ? GasCosts::txDataNonZeroGas : GasCosts::createDataGas),
+		// Data gas for data itself
+		dataGas(toCompactBigEndian(m_value, 1))
+	);
+}
+
+AssemblyItems ShortCodeCopyMethod::execute(Assembly& _assembly) const
+{
+	bytes data = toCompactBigEndian(m_value, 1);
+	assertThrow(data.size() >= 1 && data.size() <= 32, OptimizerException, "Invalid number encoding.");
+	AssemblyItems actualCopyRoutine = copyRoutine();
+	actualCopyRoutine[7] = _assembly.newData(data);
+	if (data.size() != 32)
+	{
+		// Update sizes for short encodings.
+		actualCopyRoutine[6] = u256(data.size());
+		actualCopyRoutine[8] = u256(32 - data.size());
+	}
+	return actualCopyRoutine;
+}
+
+AssemblyItems const& ShortCodeCopyMethod::copyRoutine()
+{
+	AssemblyItems static copyRoutine{
+		// constant to be reused 3+ times
+		u256(0),
+
+		// back up memory
+		// mload(0)
+		Instruction::DUP1,
+		Instruction::MLOAD,
+
+		// clear out memory (to avoid dirty upper bits)
+		Instruction::DUP1,
```

Comment:
Damn, this should be `DUP2`. Need to improve tests because none of this is detected.

Reply:
Now I feel the old pain again. I thought yul banished those wraiths...